MVP Merkle Tree (#12)
CimimUxMaio/merkle-tree: README.md

# Merkle Tree
A Merkle Tree implementation written in Rust.

# Usage
```rust
use merkle_tree::{MerkleTree, MerkleProof};

// A Merkle Tree can be built out of an array.
let tree = MerkleTree::build(&[1, 2, 3]);

// A Merkle Tree can generate a proof that it contains an element.
let proof: MerkleProof = tree.get_proof(1);

// A Merkle Tree can verify that a given hash is contained in it.

> let contains: bool = proof.verify(2); // Does the tree contain number 2?

Reviewer: - Technically speaking it seems that the verification is done in the `MerkleProof`, not in the `MerkleTree`.
- You can remove the `: bool` in line 15 because of type inference.
- It would be better if you add an assertion or a print statement after declaring the boolean. If it's a print statement you can add a comment next to it with the expected output.
- I'm not sure what value `contains` will have by reading this example. By calling `verify(2)` we are verifying that the tree contains the number 2. But we are calling `verify` from a proof generated with `get_proof(1)` so what does the argument of `get_proof` mean?

Author: I updated comments and included your suggestion.

// It is possible to create an empty tree.
let mut empty_tree = MerkleTree::build::<u8>(&[]);

// A Merkle Tree can be dynamic, this means that elements can be added once it is built.
empty_tree.push(4);
```

- You can run `make docs` to check the full documentation.

# How it Works
Leaf nodes are generated with the hashes of each input element. Then, parent nodes are computed

> as the hash between both of their child nodes until reaching the root.

Reviewer: Please add more details on how a Merkle Tree works:

- Is this a binary tree?
- What does each input element represent?

Author: I added more detail to the explanation.

To be able to handle any amount of elements, trees have an allocated capacity which will always
be a power of 2. This means that if the amount of elements from which it is being generated

> is not a power of 2, aditional "padding" values will be included until reaching the tree's capacity.

Reviewer: Does `amount of elements from which it is being generated` mean `input elements`?

Reviewer: Also there's a typo:

```suggestion
be a power of 2. This means that if the amount of elements from which it is being generated
is not a power of 2, additional "padding" values will be included until reaching the tree's capacity.
```

Author: Updated.

When a new element is pushed, if there is enough capacity to allocate it, the new value's hash will
replace the first "padding" leaf value and will trigger the update of only the parents in the

> corresponding section of the tree.

Reviewer: ```suggestion
replace the first "padding" leaf value and will only trigger the update of the parents of the updated leaf value.
```

Reviewer: or

```suggestion
replace the first "padding" leaf value and will only trigger the update of the ancestors of the updated leaf value.
```

Author: Updated.

> If there is not enough capacity, the tree's capacity will be automatically doubled in order to

Reviewer: Here it seems that `capacity` from `enough capacity` and `tree's capacity` are different. Maybe this could clarify it?

```suggestion
If there is not enough capacity, it will be automatically doubled in order to
```

Author: Updated.
